virtio: add virtiofs

fkokosinski opened the pull request:
This PR:
* adds virtiofs functions and required fuse structures
* implements zephyr filesystem operations for virtiofs
* adds virtiofs sample that presents use of this filesystem

~Note that this draft PR uses the VIRTIO PCI driver implemented in #83892, and is meant to showcase how the VIRTIO PCI driver can be used with Zephyr.~

jmichalski-ant committed:
fs: implement zephyr fs api for virtiofs

This commit implements zephyr filesystem operations for virtiofs

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

jmichalski-ant committed:
fs: virtiofs: add virtiofs sample

This commit adds virtiofs sample that presents use of this filesystem

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

cfriedt commented:
@fkokosinski -

Looks very good so far.

Is it possible to separate the virtiofs and fuse contributions into separate commits? I think they can be used independently - e.g. network based FUSE.

I was initially searching for the separation between host and device roles under `drivers/virtio`, but now it's a bit more clear that `drivers/virtio` is mainly for common components.

Maybe slightly different naming would make it easier to differentiate which is the host driver and which is the device between e.g. with virtiofs.c and virtiofs_zvfs.c, maybe virtiofs_host.c and virtiofs_device.c. Naming things is hard though, so it's maybe good to discuss. Additionally, different subsystems have different naming conventions for things, but it would be nice to have some cross-subsystem naming. 

Given that it might be hard to identify all of the different virtio drivers that will eventually be scattered throughout Zephyr, and which might do the host side vs device side, I would also like to see a new section in the documentation.

* maybe place things under `doc/hardware/virtualization`
* good high-level topics
  * Overview
  * Concepts (an SVG diagram would go a long way here)
  * Host-side drivers
  * Device-side drivers
  * Samples

Happy to help move this along any way I can.

jmichalski-ant committed:
fs: add fuse primitives

This commit adds fuse structures, requests and functions to fill them

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>

jmichalski-ant committed:
fs: add virtiofs operations

This commit adds virtiofs functions implementing fuse operations required
to enable zephyr filesystem support

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>

jmichalski-ant committed:
docs: virtio: add section about available drivers and samples

This commit adds sections to the virtio docs with the currently
supported transfer methods, drivers and samples

Signed-off-by: Jakub Michalski <jmichalski@antmicro.com>

fkokosinski commented on include/zephyr/fs/fuse.h:
```diff
@@ -0,0 +1,394 @@
+/*
+ * Copyright (c) 2025 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
We moved the code that was based on thi sheader to a separate file (`fuse_abi.h`), with the original BSD license

cfriedt replied:
Shouldn't it be possible to install this header with a kernel headers package of some kind?

Why are we even duplicating it inside of the Zephyr repo?

fkokosinski replied:
Unfortunately, there are some issues with including externally supplied headers:
* Layout of FUSE structures is not stable and was changing in the past (e.g. struct sizes were changed regularly and in some early commits even struct names and field names were sometimes altered) and presumably will also change in the future. This may cause all sorts of problems if a subset of this header used by Zephyr is modified. The virtiofsd itself doesn't use this header, instead it uses its own definitions (https://gitlab.com/virtio-fs/virtiofsd/-/blob/main/src/fuse.rs) and as of now its already a few versions behind the `fuse.h` from Linux
* Zephyr allows building on platforms different than Linux and this header won't be present on those. For example, up until recently (QEMU 8.0), virtiofsd was supporting macOS hosts (in QEMU 8.0 implementation was changed to rust based one and non-linux support was dropped), so including header from Linux would artificially limit this use case.

fkokosinski commented on samples/subsys/fs/virtiofs/README.rst:
```diff
@@ -0,0 +1,73 @@
+   virtiofsd --socket-path=/tmp/vhostqemu -o source=shared_dir_path
+
+where :code:`shared_dir_path` is a directory that will be mounted on Zephyr side.
+Then you can launch the QEMU using:
```
Fixed

fkokosinski commented on samples/subsys/fs/virtiofs/README.rst:
```diff
@@ -0,0 +1,73 @@
+
+This sample will list files and directories in the mounted filesystem and print contents of file :code:`file` in the mounted directory.
+This sample will also create some files and directories.
+You can create sample directory using :code:`prepare_sample_directory.sh`.
```
Fixed

fkokosinski commented on samples/subsys/fs/virtiofs/README.rst:
```diff
@@ -0,0 +1,73 @@
+Overview
+********
+
+This sample app demonstrates use of Zephyr's :ref:`file system API
```
Done

fkokosinski commented on subsys/fs/fuse/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2025 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+add_library(FUSE INTERFACE)
```
We've separated them after a suggestion made by @cfriedt here: https://github.com/zephyrproject-rtos/zephyr/pull/86768#issuecomment-2783560681.

You're right that this FUSE structure is upside-down. To avoid confision with typical non-inverted FUSE usage, we've renamed `fuse` to `fuse_client`

aescolar replied:
If split, would it be better to place these fuse_* headers under the subsys/fs/fuse_client instead of in the include directory? (I'm not going to argue this, or block this PR based on this. But placing them here give the impression this is a public API meant to be used by the app or other subsystems).

fkokosinski commented on include/zephyr/fs/fuse.h:
```diff
@@ -0,0 +1,394 @@
+
+struct fuse_out_header {
+	uint32_t len;
+	int32_t	error;
```
Fixed

aescolar commented on include/zephyr/fs/fuse_client.h:
```diff
@@ -0,0 +1,161 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
```
Thanks for the rename. Could you also add a file comment telling a bit what this is meant to be used for?
Maybe something in the line of "This header provides helper functions to pack FUSE client requests. Note the client is the side which initiates these requests, and that in a typical FUSE usage the client would be the Linux kernel. While in Zephyr's case this is to enable functionality like the embedded virtiofs client connecting to a virtiofsd daemon running in the host." (or whatever content you think would help a future developer understand what they are looking at).

aescolar commented on doc/hardware/virtualization/virtio.rst:
```diff
@@ -123,6 +123,31 @@ virtqueue has to be acquired using :c:func:`virtio_get_virtqueue`. To send data
+
+Virtiofs
+=========
+This driver provides support for `virtiofs <https://virtio-fs.gitlab.io/>`_ - a filesystem allowing virtual machine guest to access
```
```suggestion
This driver provides support for `virtiofs <https://virtio-fs.gitlab.io/>`_ - a filesystem allowing a virtual machine guest to access
```

fkokosinski replied:
Fixed

aescolar commented on doc/hardware/virtualization/virtio.rst:
```diff
@@ -123,6 +123,31 @@ virtqueue has to be acquired using :c:func:`virtio_get_virtqueue`. To send data
+This driver provides support for `virtiofs <https://virtio-fs.gitlab.io/>`_ - a filesystem allowing virtual machine guest to access
+a directory on the host. It uses FUSE messages to communicate between the host and the guest in order to perform filesystem operations such as
+opening and reading files. Every time the guest wants to perform some filesystem operation it places in the virtqueue a descriptor chain
+starting with device readable part, containing the FUSE input header and input data, and ending it with device writeable part, with place
```
```suggestion
starting with the device readable part, containing the FUSE input header and input data, and ending it with the device writeable part, with place
```

fkokosinski replied:
Done

aescolar commented on doc/hardware/virtualization/virtio.rst:
```diff
@@ -123,6 +123,31 @@ virtqueue has to be acquired using :c:func:`virtio_get_virtqueue`. To send data
+a directory on the host. It uses FUSE messages to communicate between the host and the guest in order to perform filesystem operations such as
+opening and reading files. Every time the guest wants to perform some filesystem operation it places in the virtqueue a descriptor chain
+starting with device readable part, containing the FUSE input header and input data, and ending it with device writeable part, with place
+for FUSE output header and output data.
```
```suggestion
for the FUSE output header and output data.
```

fkokosinski replied:
Done

aescolar commented on include/zephyr/fs/virtiofs.h:
```diff
@@ -0,0 +1,16 @@
+
+#ifndef ZEPHYR_INCLUDE_FS_VIRTIOFS_H_
+#define ZEPHYR_INCLUDE_FS_VIRTIOFS_H_
+
```
guard missing :)

aescolar reviewed (commented):
Just one trivial issue

aescolar reviewed (dismissed):
+1
note I have not gone thru the code in detail

kgugala reviewed (dismissed):
LGTM

fkokosinski commented:
> Licenses make ci red

@de-nordic right. I've tagged this PR with the `TSC` label so that it is brought to the TSC's attention, which I believe is the required next step for this PR.

In the meantime, could you please revisit this PR and let me know if there's anything else that still needs to be addressed here?

fkokosinski commented:
> > Yes, it's about this single BSD-licensed header file, which specifies the FUSE ABI: [`b6cb351`/subsys/fs/fuse_client/fuse_abi.h](https://github.com/zephyrproject-rtos/zephyr/blob/b6cb351e5418d635341b72606e45c73c4a3aa851/subsys/fs/fuse_client/fuse_abi.h).
> 
> isnt this targeting native simulator only?

We’re using this with QEMU. We’re showcasing this in the `subsys/fs/virtiofs` sample this PR is adding: https://github.com/zephyrproject-rtos/zephyr/tree/b6cb351e5418d635341b72606e45c73c4a3aa851/samples/subsys/fs/virtiofs

nashif commented:
> We’re using this with QEMU. We’re showcasing this in the `subsys/fs/virtiofs` sample this PR is adding: [`b6cb351`/samples/subsys/fs/virtiofs](https://github.com/zephyrproject-rtos/zephyr/tree/b6cb351e5418d635341b72606e45c73c4a3aa851/samples/subsys/fs/virtiofs)

ok, thanks.

nashif commented on subsys/fs/fuse_client/fuse_abi.h:
```diff
@@ -0,0 +1,269 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
```
originally in Linux this file is licensed:
```
SPDX-License-Identifier: ((GPL-2.0 WITH Linux-syscall-note) OR BSD-2-Clause) 
```

looks like you are picking BSD-2 here, the question if this is something we can actually do, because here you are basically changing the license and dropping GPL. Need to check if we should keep the original license as is.

fkokosinski replied:
Right.  I've changed the `fuse_abi.h` license so that the original identifier is preserved.
